File: static/js/prediction.js
const sum_diario_comercial = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "L'HOSPITALET LLOBR.",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB.",
  "SANT FELIU LL.",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

const sum_diario_domestic = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "LES BOTIGUES SITGES",
  "L_HOSPITALET LLOBR.",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB.",
  "SANT FELIU LL.",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

const sum_diario_industrial = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "L'HOSPITALET LLOBR.",
  "LA LLAGOSTA",
  "LES BOTIGUES SITGES",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB.",
  "SANT CUGAT",
  "SANT FELIU LL.",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

const sum_mensual_comercial = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "L'HOSPITALET LLOBR",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB",
  "SANT FELIU LL",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

const sum_mensual_domestic = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "L'HOSPITALET LLOBR",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB",
  "SANT FELIU LL",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

const sum_mensual_industrial = [
  "BADALONA",
  "BARCELONA",
  "BEGUES",
  "CASTELLDEFELS",
  "CERDANYOLA",
  "CORNELLA",
  "EL PAPIOL",
  "ESPLUGUES",
  "GAVA",
  "L'HOSPITALET LLOBR",
  "LA LLAGOSTA",
  "LES BOTIGUES SITGES",
  "MONTCADA I REIXAC",
  "MONTGAT",
  "PALLEJA",
  "SANT ADRIA",
  "SANT BOI",
  "SANT CLIMENT LLOB",
  "SANT CUGAT",
  "SANT FELIU LL",
  "SANT JOAN DESPI",
  "SANT JUST DESVERN",
  "STA.COLOMA CERVELLO",
  "STA.COLOMA GRAMENET",
  "TORRELLES LLOBREGAT",
  "VILADECANS",
];

function prediction(theform) {
  types = theform.flexRadioDefault["value"];
  zona = theform.zonaSelector["value"];

  // path = "../model/modelv4/images/" + types + "/" + zona + "/";
  if (types == "mensual") {
    path = "../model/modelv4/images/mensual";
  } else {
    path = "../model/modelv4/images/error";
  }

  // images:
  prediction_image_comercial = document.querySelector(
    "#prediction-image-comercial"
  );
  analisi_image_comercial = document.querySelector("#analisi-image-comercial");

  prediction_image_domestic = document.querySelector(
    "#prediction-image-domestic"
  );
  analisi_image_domestic = document.querySelector("#analisi-image-domestic");

  prediction_image_industrial = document.querySelector(
    "#prediction-image-industrial"
  );
  analisi_image_industrial = document.querySelector(
    "#analisi-image-industrial"
  );

  // titles:
  analisi_comercial_title = document.querySelector("#analisi-comercial-title");
  prediction_comercial_title = document.querySelector(
    "#prediction-comercial-title"
  );

  analisi_domestic_title = document.querySelector("#analisi-domestic-title");
  prediction_domestic_title = document.querySelector(
    "#prediction-domestic-title"
  );

  analisi_industrial_title = document.querySelector(
    "#analisi-industrial-title"
  );
  prediction_industrial_title = document.querySelector(
    "#prediction-industiral-title"
  );

  analisi_title_area_comercial = document.querySelector("#analisi-title-area");
  analisi_image_area_comercial = document.querySelector("#analisi-image-area");
  analisi_title_area_domestico = document.querySelector(
    "#analisi-title-domestico"
  );
  analisi_image_area_domestico = document.querySelector(
    "#analisi-image-domestico"
  );
  analisi_title_area_industrial = document.querySelector(
    "#analisi-title-industrial"
  );
  analisi_image_area_industrial = document.querySelector(
    "#analisi-image-industrial"
  );

  error_1 = document.querySelector("#error-1");
  error_2 = document.querySelector("#error-2");
  error_3 = document.querySelector("#error-3");

  if (types == "diario") {
    error_1.style.display = "block";
    error_2.style.display = "block";
    error_3.style.display = "block";

    prediction_comercial_title.innerHTML =
      "Predicción del consumo diario comercial";
    prediction_domestic_title.innerHTML =
      "Predicción del consumo diario domestic";
    prediction_industrial_title.innerHTML =
      "Predicción del consumo diario industrial";
    analisi_comercial_title.innerHTML =
      "Analisi de la tendencia del consumo diario comercial";
    analisi_domestic_title.innerHTML =
      "Analisi de la tendencia del consumo diario domestico";
    analisi_industrial_title.innerHTML =
      "Analisi de la tendencia del consumo diario industria";

    if (sum_diario_comercial.includes(zona)) {
      prediction_image_comercial.src =
        path + "/prediction_comercial_" + zona + ".jpeg";
      analisi_image_comercial.src = analisi_image_comercial =
        path + "/param_comercial_" + zona + ".jpeg";
    } else {
      prediction_image_comercial.src = "../static/update/prediction.png";
      analisi_image_comercial.src = "../static/update/analisi.png";
    }

    if (sum_diario_domestic.includes(zona)) {
      prediction_image_domestic.src =
        path + "/prediction_domestic_" + zona + ".jpeg";
      analisi_image_domestic.src = path + "/param_domestic_" + zona + ".jpeg";
    } else {
      prediction_image_domestic.src = "../static/update/prediction.png";
      analisi_image_domestic.src = "../static/update/analisi.png";
    }

    if (sum_diario_industrial.includes(zona)) {
      prediction_image_industrial.src =
        path + "/prediction_industrial_" + zona + ".jpeg";
      analisi_image_industrial.src =
        path + "/param_industrial_" + zona + ".jpeg";
    } else {
      prediction_image_industrial.src = "../static/update/prediction.png";
      analisi_image_industrial.src = "../static/update/analisi.png";
    }
  } else if (types == "mensual") {
    error_1.style.display = "block";
    error_2.style.display = "block";
    error_3.style.display = "block";

    prediction_comercial_title.innerHTML =
      "Predicción del consumo mensual comercial";
    prediction_domestic_title.innerHTML =
      "Predicción del consumo mensual domestic";
    prediction_industrial_title.innerHTML =
      "Predicción del consumo mensual industrial";
    analisi_comercial_title.innerHTML =
      "Analisi de la tendencia del consumo mensual comercial";
    analisi_domestic_title.innerHTML =
      "Analisi de la tendencia del consumo mensual domestico";
    analisi_industrial_title.innerHTML =
      "Analisi de la tendencia del consumo mensual industria";

    if (sum_diario_comercial.includes(zona)) {
      prediction_image_comercial.src =
        path + "/prediction_comercial_" + zona + ".jpeg";
      analisi_image_comercial.src = analisi_image_comercial =
        path + "/param_comercial_" + zona + ".jpeg";
    } else {
      prediction_image_comercial.src = "../static/update/prediction.png";
      analisi_image_comercial.src = "../static/update/analisi.png";
    }

    if (sum_diario_domestic.includes(zona)) {
      prediction_image_domestic.src =
        path + "/prediction_domestic_" + zona + ".jpeg";
      analisi_image_domestic.src = path + "/param_domestic_" + zona + ".jpeg";
    } else {
      prediction_image_domestic.src = "../static/update/prediction.png";
      analisi_image_domestic.src = "../static/update/analisi.png";
    }

    if (sum_diario_industrial.includes(zona)) {
      prediction_image_industrial.src =
        path + "/prediction_industrial_" + zona + ".jpeg";
      analisi_image_industrial.src =
        path + "/param_industrial_" + zona + ".jpeg";
    } else {
      prediction_image_industrial.src = "../static/update/prediction.png";
      analisi_image_industrial.src = "../static/update/analisi.png";
    }
  } else if (types == "error") {
    prediction_comercial_title.innerHTML =
      "Detecció de errores del consumo diario comercial";
    prediction_domestic_title.innerHTML =
      "Detecció de errores del consumo diario domestic";
    prediction_industrial_title.innerHTML =
      "Detecció de errores del consumo diario industrial";

    error_1.style.display = "none";
    error_2.style.display = "none";
    error_3.style.display = "none";

    if (sum_diario_comercial.includes(zona)) {
      prediction_image_comercial.src =
        path + "/error_comercial_" + zona + ".jpeg";
    } else {
      prediction_image_comercial.src = "../static/update/prediction.png";
    }

    if (sum_diario_domestic.includes(zona)) {
      prediction_image_domestic.src =
        path + "/error_domestic_" + zona + ".jpeg";
    } else {
      prediction_image_domestic.src = "../static/update/prediction.png";
    }

    if (sum_diario_industrial.includes(zona)) {
      prediction_image_industrial.src =
        path + "/error_industrial_" + zona + ".jpeg";
    } else {
      prediction_image_industrial.src = "../static/update/prediction.png";
    }
  } else {
    prediction_image_comercial.src = "../static/update/prediction.png";
    analisi_image_comercial.src = "../static/update/analisi.png";
    prediction_image_domestic.src = "../static/update/prediction.png";
    analisi_image_domestic.src = "../static/update/analisi.png";
    prediction_image_industrial.src = "../static/update/prediction.png";
    analisi_image_industrial.src = "../static/update/analisi.png";
  }
}

File: model/modelv2/prediction.py
#pip install neuralprophet

import os
import plotly.graph_objects as go
import numpy as np
import pandas as pd
import torch
import kaleido
from neuralprophet import NeuralProphet


def ARnet(data_location,
          section2,
          period=4,
          save_prediction=False,
          save_img=False,
          save_model=False,
          show=False):

    section1 = pd.read_excel(data_location)
    consumo = data_location.split("\\")[-1][:-3]
    mensual = consumo.split("_")[1] == "mensual"
    actividad = consumo.split("_")[2][:-2]

    if mensual == True:
        df = pd.DataFrame({"ds": section1.iloc[:, 1], "y": section1[section2]})
        m = NeuralProphet(loss_func=torch.nn.L1Loss)
        df_train, df_test = m.split_df(df, freq="M", valid_p=0.1)
        metrics = m.fit(df_train, freq="M", validation_df=df_test)
    else:
        df = pd.DataFrame({"ds": section1["FECHA"], "y": section1[section2]})
        m = NeuralProphet(loss_func=torch.nn.L1Loss)
        df_train, df_test = m.split_df(df, freq="D", valid_p=0.1)
        metrics = m.fit(df_train, freq="D", validation_df=df_test)

    cost = metrics.loc[metrics.shape[0] - 1, "RMSE"]
    pred_train = m.predict(df)
    next_dataset = m.make_future_dataframe(df_test, periods=period)
    pred_next = m.predict(df=next_dataset)
    df = pd.concat([pred_train, pred_next])

    if save_prediction == True:
        path = "./model/modelv2/prediction/"
        if not os.path.exists(path):
            os.makedirs(path)
        df.to_csv(path + f"result_{actividad}_{section2}.csv")

    if mensual == True:
        final = go.Figure()
        final.add_trace(
            go.Scatter(x=df['ds'],
                       y=df["yhat1"],
                       mode='lines',
                       name='Training set'))
        final.update_layout(title={
            'text':
            f"Predicción del consumo mensual(Litro/Mes) {actividad} en {section2}",
            'y': 0.95,
            'x': 0.5,
            'xanchor': 'center',
            'yanchor': 'top',
            'font': {
                'size': 30
            },
        },
                            legend={
                                "font": {
                                    "size": 18
                                },
                            },
                            xaxis={
                                "tickfont": {
                                    "size": 22
                                },
                                "title": {
                                    "text": "Meses",
                                    "font": {
                                        "size": 24
                                    },
                                },
                            },
                            yaxis={
                                "tickfont": {
                                    "size": 22
                                },
                                "title": {
                                    "text": "Consumo(Litro/Día)",
                                    "font": {
                                        "size": 24
                                    },
                                },
                            })
        final.add_trace(
            go.Scatter(x=df['ds'][35:35 + period + 1],
                       y=df['yhat1'][35:35 + period + 1],
                       mode='lines',
                       name='Prediction'))
        final.add_trace(
            go.Scatter(x=df.loc[:34, 'ds'],
                       y=df.loc[:34, 'y'],
                       mode='markers',
                       name='Actual',
                       line_color="firebrick"))

        fig_param = m.plot_parameters(plotting_backend="plotly")
        fig_param.update_layout(
            title={
                'text':
                f"Análisis de la tendencia del consumo {actividad} en {section2}",
                'y': 0.99,
                'x': 0.5,
                'xanchor': 'center',
                'yanchor': 'top',
                'font': {
                    'size': 16
                },
            })

    else:
        final = go.Figure()
        final.add_trace(
            go.Scatter(x=df['ds'],
                       y=df["yhat1"],
                       mode='lines',
                       name='Training set'))
        final.update_layout(title={
            'text':
            f"Predicción del consumo diario(Litro/Día) {actividad} en {section2}",
            'y': 0.95,
            'x': 0.5,
            'xanchor': 'center',
            'yanchor': 'top',
            'font': {
                'size': 30
            },
        },
                            legend={
                                "font": {
                                    "size": 18
                                },
                            },
                            xaxis={
                                "tickfont": {
                                    "size": 22
                                },
                                "title": {
                                    "text": "Meses",
                                    "font": {
                                        "size": 24
                                    },
                                },
                            },
                            yaxis={
                                "tickfont": {
                                    "size": 22
                                },
                                "title": {
                                    "text": "Consumo(Litro/Día)",
                                    "font": {
                                        "size": 24
                                    },
                                },
                            })
        final.add_trace(
            go.Scatter(x=df['ds'][1078:1078 + period + 1],
                       y=df['yhat1'][1078:1078 + period + 1],
                       mode='lines',
                       name='Prediction'))
        final.add_trace(
            go.Scatter(x=df['ds'],
                       y=df['y'],
                       mode='markers',
                       name='Actual',
                       line_color="firebrick"))

        fig_param = m.plot_parameters(plotting_backend="plotly")
        fig_param.update_layout(
            title={
                'text':
                f"Análisis de la tendencia del consumo {actividad} en {section2}",
                'y': 0.99,
                'x': 0.5,
                'xanchor': 'center',
                'yanchor': 'top',
                'font': {
                    'size': 16
                },
            })

    final.add_annotation(text=f"RMSE: {round(cost,2)}",
                         xref="paper",
                         yref="paper",
                         x=0.9,
                         y=0.1)

    if save_img == True:

        path = "./model/modelv2/images/"
        if mensual == True:
            path += "mensual/" + section2 + "/"
        else:
            path += "diario/" + section2 + "/"

        if not os.path.exists(path):
            os.makedirs(path)

        fig_param.write_image(path + f"param_{actividad}_{section2}.jpeg")
        final.write_image(path + f"prediction_{actividad}_{section2}.jpeg",
                          width=1980,
                          height=1080)

    if show == True:
        fig_param.show()
        final.show()

    if save_model == True:
        #####Saves the model's weights and state in the file model.pth, which allows the continuous update once it is deployed on the website.

        path = "./model/modelv2/save_model/"
        if not os.path.exists(path):
            os.makedirs(path)

        torch.save(m.model.state_dict(),
                   path + f"model_{actividad}_{section2}.pth")

    return m, pred_train, pred_next, final, fig_param, metrics  #pred_train -> train + test


data_location = [
    "./model/modelv2/sum_mensual_comercial.xlsx"
    "./model/modelv2/data/DOMESTICO_NORM_resum.xlsx",
    "./model/modelv2/data/INDUS_NORM_resum.xlsx"
    "./model/modelv2/sum_diario_comercial.xlsx"
]

zonas = [
    "BADALONA", "BARCELONA", "BEGUES", "CASTELLDEFELS", "CERDANYOLA",
    "CORNELLA", "EL PAPIOL", "ESPLUGUES", "GAVA", "L'HOSPITALET LLOBR.",
    "LA LLAGOSTA", "LES BOTIGUES SITGES", "MONTCADA I REIXAC", "MONTGAT",
    "PALLEJA", "SANT ADRIA", "SANT BOI", "SANT CLIMENT LLOB.", "SANT CUGAT",
    "SANT FELIU LL.", "SANT JOAN DESPI", "SANT JUST DESVERN",
    "STA.COLOMA CERVELLO", "STA.COLOMA GRAMENET", "TORRELLES LLOBREGAT",
    "VILADECANS"
]

if __name__ == "__main__":
    print("prediction model:\n")

    ###Noted that the last available data month(2021 December) is not complete, which shows a significant decay in the representation.
    data_location = "./model/modelv2/data/sum_mensual_comercial.xlsx"
    model = ARnet(data_location,
                  "BARCELONA",
                  save_prediction=True,
                  save_img=True,
                  save_model=True,
                  show=False)

    ###Noted that the last available data month(2021 December) is not complete, which shows a significant decay in the representation.
    # data_location = "./model/modelv2/data/sum_diario_comercial.xlsx"
    # model = ARnet(data_location,
    #               "BARCELONA",
    #               save_prediction=True,
    #               save_img=True,
    #               save_model=True,
    #               show=False)
